pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go

```diff
@@ -308,6 +333,43 @@ func updateStatus(gatewayInstance *mesh_k8s.GatewayInstance, svc *kube_core.Serv
 	}
 }
 
+const serviceKey string = ".metadata.service"
+
+func GatewayToInstanceMapper(l logr.Logger, client kube_client.Client) kube_handler.MapFunc {
+	l = l.WithName("gateway-to-gateway-instance-mapper")
+
+	return func(obj kube_client.Object) []kube_reconile.Request {
+		gateway := obj.(*mesh_k8s.Gateway)
+
+		// Take first service from selectors
+		var serviceName string
+		for _, selector := range gateway.Spec.GetSelectors() {
+			tagValue, ok := selector.Match[mesh_proto.ServiceTag]
+			if ok {
+				serviceName = tagValue
+				break
+			}
+		}
+
+		instances := &mesh_k8s.GatewayInstanceList{}
+		if err := client.List(
+			context.Background(), instances, kube_client.MatchingFields{serviceKey: serviceName},
+		); err != nil {
+			l.WithValues("gateway", obj.GetName()).Error(err, "failed to fetch GatewayInstances")
+			return nil
+		}
+
+		var req []kube_reconile.Request
+		for _, instance := range instances.Items {
+			req = append(req, kube_reconile.Request{
+				NamespacedName: kube_types.NamespacedName{Namespace: instance.Namespace, Name: instance.Name},
```

Comment:
What would happen when a Gateway changes its service tags such that it unbinds from a GatewayInstance?

Reply:
[`controller-runtime enqueues both the old and the new object for an Update`](https://github.com/kubernetes-sigs/controller-runtime/blob/f236f0345ad2933912ebf34bfcf0f93620769654/pkg/handler/enqueue_mapped.go#L63-L64) so the unbound-from `GatewayInstance` will be enqueued.

Reply:
OK, so `GatewayToInstanceMapper` will get called twice with the old and new objects. When it gets called with the old, then we should find the old binding and enqueue and update for that, and when it gets called with the new we would find the new binding. The final state would then be updated my the main reconciler.

If that's correct, I think it's worth leaving a comment.